File: frontend/src/lib/book.ts
import { api, AuthToken, handleError, setAuthHeader } from "@/lib/api";

interface IndustryIdentifier {
  type: string;
  identifier: string;
}

interface ReadingModes {
  text: boolean;
  image: boolean;
}

interface PenalizationSummary {
  containsEpubBubbles: boolean;
  containsImageBubbles: boolean;
}

interface ImageLinks {
  smallThumbnail: string;
  thumbnail: string;
}

interface VolumeInfo {
  title: string;
  subtitle?: string;
  authors?: string[];
  publisher?: string;
  publishedDate: string;
  description?: string;
  industryIdentifiers: IndustryIdentifier[];
  readingModes: ReadingModes;
  pageCount: number;
  printType: string;
  categories: string[];
  maturityRating: string;
  allowAnonLogging: boolean;
  contentVersion: string;
  panelizationSummary: PenalizationSummary;
  imageLinks: ImageLinks;
  language: string;
  previewLink: string;
  infoLink: string;
  canonicalVolumeLink: string;
  averageRating?: number;
  ratingsCount?: number;
}

interface ListPrice {
  amount: number;
  currencyCode: string;
}

interface RetailPrice {
  amount: number;
  currencyCode: string;
}

interface ListPrice1 {
  amountInMicros: number;
  currencyCode: string;
}

interface RetailPrice1 {
  amountInMicros: number;
  currencyCode: string;
}

interface RentalDuration {
  unit: string;
  count: number;
}

interface Offer {
  finskyOfferType: number;
  listPrice: ListPrice1;
  retailPrice: RetailPrice1;
  giftable?: boolean;
  rentalDuration?: RentalDuration;
}

interface SaleInfo {
  country: string;
  saleability: string;
  isEbook: boolean;
  listPrice?: ListPrice;
  retailPrice?: RetailPrice;
  buyLink?: string;
  offers?: Offer[];
}

interface Epub {
  isAvailable: boolean;
  acsTokenLink?: string;
  downloadLink?: string;
}

interface Pdf {
  isAvailable: boolean;
  acsTokenLink?: string;
}

interface AccessInfo {
  country: string;
  viewability: string;
  embeddable: boolean;
  publicDomain: boolean;
  textToSpeechPermission: string;
  epub: Epub;
  pdf: Pdf;
  webReaderLink: string;
  accessViewStatus: string;
  quoteSharingAllowed: boolean;
}

interface SearchInfo {
  textSnippet: string;
}

export interface BookVolume {
  kind?: string; // Default values are not directly supported in TypeScript interfaces, consider initializing in class constructor if using classes.
  id?: string;
  etag?: string;
  selfLink?: string;
  volumeInfo?: VolumeInfo;
  saleInfo?: SaleInfo;
  accessInfo?: AccessInfo;
  searchInfo?: SearchInfo;
}

export interface BookSearchResponse {
  kind?: string;
  totalItems: number;
  items: BookVolume[];
}


interface BookSearchData extends AuthToken {
  q: string;
  page: string;
}

interface BookData extends AuthToken {
  id: string;
}

export async function searchBooks({ q, page, accessToken }: BookSearchData): Promise<BookSearchResponse> {
  const params = new URLSearchParams({ q, page });
  const response = await api.get(`/book/search?${params.toString()}`, setAuthHeader(accessToken));
  return handleError(response);
}


export async function get({ id, accessToken }: BookData): Promise<BookVolume> {
  const response = await api.get(`/book/${id}`, setAuthHeader(accessToken));
  return handleError(response);
}

File: backend/book/schemas/book.py
from typing import List, Optional

from pydantic import BaseModel


class IndustryIdentifier(BaseModel):
    type: str
    identifier: str


class ReadingModes(BaseModel):
    text: bool
    image: bool


class PanelizationSummary(BaseModel):
    containsEpubBubbles: bool
    containsImageBubbles: bool


class ImageLinks(BaseModel):
    smallThumbnail: str
    thumbnail: str


class VolumeInfo(BaseModel):
    title: str = ""
    subtitle: Optional[str] = None
    authors: Optional[List[str]] = None
    publisher: Optional[str] = None
    publishedDate: str = ""
    description: Optional[str] = None
    industryIdentifiers: Optional[List[IndustryIdentifier]] = None
    readingModes: ReadingModes
    pageCount: int = 0
    printType: str = ""
    categories: Optional[List[str]] = None
    maturityRating: str = ""
    allowAnonLogging: bool = False
    contentVersion: str = ""
    panelizationSummary: Optional[PanelizationSummary] = None
    imageLinks: Optional[ImageLinks] = None
    language: str = ""
    previewLink: str = ""
    infoLink: str = ""
    canonicalVolumeLink: str = ""
    averageRating: Optional[float] = None
    ratingsCount: Optional[int] = None


class ListPrice(BaseModel):
    amount: float
    currencyCode: str


class RetailPrice(BaseModel):
    amount: float
    currencyCode: str


class ListPrice1(BaseModel):
    amountInMicros: int
    currencyCode: str


class RetailPrice1(BaseModel):
    amountInMicros: int
    currencyCode: str


class RentalDuration(BaseModel):
    unit: str
    count: int


class Offer(BaseModel):
    finskyOfferType: int
    listPrice: ListPrice1
    retailPrice: RetailPrice1
    giftable: Optional[bool] = None
    rentalDuration: Optional[RentalDuration] = None


class SaleInfo(BaseModel):
    country: str
    saleability: str
    isEbook: bool
    listPrice: Optional[ListPrice] = None
    retailPrice: Optional[RetailPrice] = None
    buyLink: Optional[str] = None
    offers: Optional[List[Offer]] = None


class Epub(BaseModel):
    isAvailable: bool
    acsTokenLink: Optional[str] = None
    downloadLink: Optional[str] = None


class Pdf(BaseModel):
    isAvailable: bool
    acsTokenLink: Optional[str] = None


class AccessInfo(BaseModel):
    country: str
    viewability: str
    embeddable: bool
    publicDomain: bool
    textToSpeechPermission: str
    epub: Epub
    pdf: Pdf
    webReaderLink: str
    accessViewStatus: str
    quoteSharingAllowed: bool


class SearchInfo(BaseModel):
    textSnippet: str


class BookVolume(BaseModel):
    kind: str = ""
    id: str = ""
    etag: str = ""
    selfLink: str = ""
    volumeInfo: Optional[VolumeInfo] = None
    saleInfo: Optional[SaleInfo] = None
    accessInfo: Optional[AccessInfo] = None
    searchInfo: Optional[SearchInfo] = None


class BookSearchResponse(BaseModel):
    kind: Optional[str] = None
    totalItems: Optional[int] = None
    items: Optional[List[BookVolume]] = None
